servicemanager/servicemanager.go

```diff
@@ -0,0 +1,260 @@
+		},
+	)
+}
+
+func (sm *ServiceManager) GetServiceInfo(serviceID string) (serviceInfo ServiceInfo, err error) {
+	serviceInfo, err = sm.serviceInfoProvider.GetService(serviceID)
+	return serviceInfo, aoserrors.Wrap(err)
+}
+
+func (sm *ServiceManager) GetImageParts(service ServiceInfo) (parts ImageParts, err error) {
+	return getImageParts(service.ImagePath)
+}
+
+/**********************************************************************************************************************
+ * Private
+ *********************************************************************************************************************/
+
+func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL string, fileInfo image.FileInfo) error {
+	serviceFromStorage, err := sm.serviceInfoProvider.GetService(newService.ServiceID)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	var isServiceExist bool
+
+	if err == nil {
+		isServiceExist = true
+
+		if newService.AosVersion <= serviceFromStorage.AosVersion {
+			return ErrVersionMismatch
+		}
+	}
+
+	// download and unpack
+	urlVal, err := url.Parse(imageURL)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	var serviceImage string
+
+	if urlVal.Scheme != "file" {
+		if serviceImage, err = image.Download(context.Background(), sm.downloadDir, imageURL); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		defer os.RemoveAll(serviceImage)
+	} else {
+		serviceImage = urlVal.Path
+	}
+
+	if err = image.CheckFileInfo(context.Background(), serviceImage, fileInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// create install dir
+	log.WithFields(log.Fields{"serviceID": newService.ServiceID}).Debug("Install service")
+
+	newService.ImagePath, err = ioutil.TempDir(sm.servicesDir, "")
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	defer func() {
+		if err != nil {
+			_ = os.RemoveAll(newService.ImagePath)
+		}
+	}()
+
+	if err = imageutils.UnpackTarImage(serviceImage, newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = validateUnpackedImage(newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if isServiceExist {
+		newService.GID = serviceFromStorage.GID
+	} else {
+		newService.GID, err = sm.gidPool.GetFreeID()
+		if err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	if err = sm.prepareServiceFS(newService.ImagePath, newService.GID); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	newService.ManifestDigest, err = getManifestChecksum(newService.ImagePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = sm.serviceInfoProvider.AddService(newService); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) prepareServiceFS(imagePath string, gid int) (err error) {
+	imageParts, err := getImageParts(imagePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	tmpTar := path.Join(imagePath, "tmpTar")
```

Comment:
Add const for tmp file name. The name should have appropriate extension: .tar or .tar.gz etc.

Reply:
done